net_connection/src/json_protocol.rs

```diff
@@ -102,75 +100,189 @@ pub struct FailureResultData {
+    // DhtDataRequestData
+    #[serde(rename = "_id")]
+    pub request_id: String,
+    #[serde(rename = "dnaAddress")]
+    pub dna_address: Address,
+    #[serde(rename = "address")]
+    pub data_address: Address,
 }
 
+/// Generic DHT data message
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq, DefaultJson, Default)]
 pub struct DhtData {
+    #[serde(rename = "dnaAddress")]
+    pub dna_address: Address,
+
+    #[serde(rename = "agentId")]
+    pub provider_agent_id: String,
+
+    #[serde(rename = "address")]
+    pub data_address: Address,
+
+    #[serde(rename = "content")]
+    pub data_content: serde_json::Value,
+}
+
+/// DHT data response from a request
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, DefaultJson, Default)]
+pub struct HandleDhtResultData {
     #[serde(rename = "_id")]
-    pub msg_id: String,
+    pub request_id: String,
+    #[serde(rename = "requesterAgentId")]
+    pub requester_agent_id: String,
 
+    // DhtData
     #[serde(rename = "dnaAddress")]
     pub dna_address: Address,
-
     #[serde(rename = "agentId")]
-    pub agent_id: String,
+    pub provider_agent_id: String,
+    #[serde(rename = "address")]
+    pub data_address: Address,
+    #[serde(rename = "content")]
+    pub data_content: serde_json::Value,
+}
 
-    pub address: String,
-    pub content: serde_json::Value,
+//--------------------------------------------------------------------------------------------------
+// DHT metadata
+//--------------------------------------------------------------------------------------------------
+
+/// Data Request from own p2p-module
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, DefaultJson)]
+pub struct DhtMetaRequestData {
+    pub attribute: String,
+
+    // DhtDataRequestData
+    #[serde(rename = "_id")]
+    pub request_id: String,
+    #[serde(rename = "dnaAddress")]
+    pub dna_address: Address,
+    #[serde(rename = "address")]
+    pub data_address: Address,
 }
 
+/// Metadata Request from another agent
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq, DefaultJson)]
-pub struct GetDhtMetaData {
+pub struct FetchDhtMetaData {
+    pub attribute: String,
+
+    // FetchDhtData
+    #[serde(rename = "fromAgentId")]
+    pub requester_agent_id: String,
     #[serde(rename = "_id")]
-    pub msg_id: String,
+    pub request_id: String,
+    #[serde(rename = "dnaAddress")]
+    pub dna_address: Address,
+    #[serde(rename = "address")]
+    pub data_address: Address,
+}
 
+/// Generic DHT metadata message
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, DefaultJson)]
+pub struct DhtMetaData {
     #[serde(rename = "dnaAddress")]
     pub dna_address: Address,
 
     #[serde(rename = "fromAgentId")]
-    pub from_agent_id: String,
+    pub provider_agent_id: String,
 
-    pub address: String,
+    #[serde(rename = "address")]
+    pub data_address: Address,
     pub attribute: String,
+    pub content: serde_json::Value,
```

Comment:
I wish there was a way these could be define in-line with the enum so we could more easily see what was going on...

I know you can do struct-style enums, but they are not very ergonomic to work with, and cause some serialization problems...

Reply:
I'm having a hard time discerning what need you are expressing here.
The github diff might be confusing, as I actually did not add this 'content' field but moved it around a bit.
I went with the idea that the `*Data` struct are reusable and not make one struct per enum variant. I see how can this add a bit of confusion but in my mind it outweighs the tedious duplication of the same fields in every struct if we did one struct per enum variant.

Reply:
I want to be able to:

```rust
pub enum MyEnum {
  pub VarOne {
    pub prop_one: String,
    pub prop_i_dont_care: String,
  }
}
```

without having to match painfully like:

```rust
match v {
  VarOne { prop_one, prop_i_dont_care: _ } => { /* ... */ },
}
```

or whatever the exact syntax is